Offline/OfflinePlayer/objc/OfflinePlayer/VideosViewController.m

```diff
@@ -0,0 +1,1370 @@
+NSString * const kDynamicDeliveryAccountID = @"5434391461001";
+NSString * const kDynamicDeliveryPolicyKey = @"BCpkADawqM0T8lW3nMChuAbrcunBBHmh4YkNl5e6ZrKQwPiK_Y83RAOF4DP5tyBF_ONBVgrEjqW6fbV0nKRuHvjRU3E8jdT9WMTOXfJODoPML6NUDCYTwTHxtNlr5YdyGYaCPLhMUZ3Xu61L";
+NSString * const kDynamicDeliveryPlaylistRefID = @"brightcove-native-sdk-plist";
+
+// For quick access to this controller from other tabs
+VideosViewController *gVideosViewController;
+
+// The Videos View Controller displays a list of HLS videos retrieved
+// from a Brightcove Dynamic Delivery account playlist.
+// You can tap the download button on a video to begin downloading the video.
+@interface VideosViewController () <BCOVPlaybackControllerDelegate, BCOVPUIPlayerViewDelegate, UITableViewDataSource, UITableViewDelegate>
+
+// Brightcove-related objects
+@property (nonatomic, strong) BCOVPlaybackService *playbackService;
+@property (nonatomic, strong) BCOVPUIPlayerView *playerView;
+@property (nonatomic, strong) BCOVFPSBrightcoveAuthProxy *authProxy;
+
+// View that holds the PlayerUI content where the video and controls are displayed
+@property (nonatomic, weak) IBOutlet UIView *videoContainer;
+
+// Keep track of info from the playlist for easy display in the table
+@property NSMutableArray<BCOVVideo *> *currentVideos;
+
+// The download queue.
+// Videos go into the preload queue first.
+// When all preloads are done, videos move to the download queue.
+@property NSMutableArray<NSDictionary *> *videoPreloadQueue;
+@property NSMutableArray<NSDictionary *> *videoDownloadQueue;
+
+// Keep track of info from the playlist
+// for easy display in the table
+@property (nonatomic, strong) NSString *currentPlaylistTitle;
+@property (nonatomic, strong) NSString *currentPlaylistDescription;
+@property (nonatomic, strong) NSMutableDictionary *imageCacheDictionary;
+
+// Table view displaying available videos from playlist, and its refresh control
+@property (nonatomic, strong) IBOutlet UITableView *videosTableView;
+@property (nonatomic, strong) UIRefreshControl *refreshControl;
+@end
+
+
+@implementation VideosViewController
+
+- (void)viewWillAppear:(BOOL)animated
+{
+    [super viewWillAppear:animated];
+
+    // Become delegate so we can control orientation
+    gVideosViewController.tabBarController.delegate = self;
+
+    [self updateStatus];
+}
+
+- (void)viewDidAppear:(BOOL)animated
+{
+    [super viewDidAppear:animated];
+    
+    [self showWarnings];
+}
+
+- (void)setup
+{
+    [self retrievePlaylist];
+    [self createNewPlaybackController];
+    [self updateStatus];
+
+    self.videoPreloadQueue = NSMutableArray.array;
+    self.videoDownloadQueue = NSMutableArray.array;
+}
+
+- (void)runPreloadVideoQueue
+{
+    NSDictionary *videoDownloadDictionary = self.videoPreloadQueue.firstObject;
+    BCOVVideo *video = videoDownloadDictionary[@"video"];
+    NSDictionary *parameters = videoDownloadDictionary[@"parameters"];
+    
+    if (video == nil)
```

Comment:
Add comment: "// Once the preload queue is empty, start the download queue"

Reply:
√